### PR TITLE
Starting the wrappers implementations with a new API

### PR DESCRIPTION
**Pull request by ffelten**
This new API aims at being more extensible (e.g. wrapping the env state to add more information such as the LogEnv wrapper). 

By returning the obs, rewards, term, trunc, info explicitly at each step, it allows to avoid playing with the state of the env to add additional information. This makes the code easier to extend.

I modified the catch example so there is a minimal example of code running with the wrappers.

Some of my changes:
- The new API returns the obs, rewards, term, trunc, info, state as a tuple
- Introduced some wrappers (see usage in the catch env)
- `VecEnv` -> automated wrapping to vectorize the environments -> autovmap
- `LogWrapper` -> automatically keeps track of the episode lengths and team reward for plotting while learning
- `AddIDToObs` -> concatenate the agent ID as a one-hot encoded vector to the obs, this allows the neural network to know who is currently playing

Next challenges:
- `AutoReset` as a wrapper: I think it is doable to actually have a wrapper doing the autoreset logic. It has been in Gymnasium and in Brax
- `NormalizeReward` & `NormalizeObs`: Algorithms often require normalization to work better. While I think normalizing the obs is not very hard (we have the bounds in the observation space), normalizing the rewards will probably require some moving avg tricks.

### CODE REVIEW

File: crazy_rl/multi_agent/jax/base_parallel_env.py [hunk 1]
```diff
@@ -26,11 +27,6 @@ class State:
     agents_locations: jnp.ndarray  # a 2D array containing x,y,z coordinates of each agent, indexed from 0.
     timestep: int  # represents the number of steps already done in the game.
 
-    observations: jnp.ndarray  # array containing the current observation of each agent.
-    rewards: jnp.ndarray  # array containing the current reward of each agent.
-    terminations: jnp.ndarray  # array of booleans which are True if the agents have crashed.
-    truncations: jnp.ndarray  # array of booleans which are True if the game reaches enough timesteps and ends.
-
 
 class BaseParallelEnv:
```
Comment (ffelten) [inline crazy_rl/multi_agent/jax/base_parallel_env.py:31]: Doc is outdated

File: crazy_rl/multi_agent/jax/catch/catch.py [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
 import jax.numpy as jnp
 import jax_dataclasses as jdc
-import numpy as np
-from jax import jit, random, vmap
+from jax import jit, random
 
 from crazy_rl.multi_agent.jax.base_parallel_env import BaseParallelEnv, State
 from crazy_rl.utils.jax_spaces import Box, Space
+from crazy_rl.utils.jax_wrappers import (
+    AddIDToObs,
```
Comment (ffelten) [inline crazy_rl/multi_agent/jax/catch/catch.py:15]: Looks weird to me to have to import all wrappers here?

File: crazy_rl/multi_agent/jax/catch/catch.py [hunk 3]
```diff
@@ -226,83 +180,49 @@ def state(self, state: State) -> jnp.ndarray:
...
-            i: number of the iteration.
-            states_key: a tuple containing states and key.
-        """
-        states, key = states_key
-
-        key, *subkeys = random.split(key, parallel_env.num_drones + 1)
+    key, *subkeys = random.split(key, num_envs + 1)
+
+    # Wrappers
+    env = NormalizeObservation(env)
```
Comment (ffelten) [inline crazy_rl/multi_agent/jax/catch/catch.py:201]: I think we shouldn't show all the wrappers here. VecEnv and AutoReset maybe but not the others.

File: crazy_rl/test/test_numpy_and_jax.py [hunk 4]
```diff
@@ -6,26 +6,26 @@
...
+    assert jax_term[1] == np_term["agent_1"]
 
-    assert state.truncations[0] == truncations["agent_0"]
-    assert state.truncations[1] == truncations["agent_1"]
+    assert jax_trunc[0] == np_trunc["agent_0"]
+    assert jax_trunc[1] == np_trunc["agent_1"]
 
-    assert state.rewards[0] > rewards["agent_0"] - 0.01
-    assert state.rewards[0] < rewards["agent_0"] + 0.01
+    assert jax_rewards[0] > np_rewards["agent_0"] - 0.01
```
Comment (ffelten) [inline crazy_rl/test/test_numpy_and_jax.py:24]: what does this test?

File: crazy_rl/utils/jax_wrappers.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        return self._env.state(state.env_state)
+
+
+class NormalizeObservation(Wrapper):
+    """Normalize the observation.
+
+    Taken and adapted from https://github.com/luchris429/purejaxrl/blob/main/purejaxrl/wrappers.py
+    """
+
+    def __init__(self, env):
```
Comment (ffelten) [inline crazy_rl/utils/jax_wrappers.py:220]: Do you think we can take the scaling range as parameters with default values being -1 and 1? This would allow to scale to any range.

File: crazy_rl/utils/jax_wrappers.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        super().__init__(env)
+
+    def reset(self, key):
+        obs, info, state = self._env.reset(key)
+
+        obs = obs / self._env.observation_space(0).high
```
Comment (ffelten) [inline crazy_rl/utils/jax_wrappers.py:226]: I guess you forgot this

File: crazy_rl/utils/jax_wrappers.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        return obs, reward / jnp.sqrt(state.var + 1e-8), term, truncated, info, state
+
+    def state(self, state: NormalizeVecRewEnvState) -> chex.Array:
+        return self._env.state(state.env_state)
+
+
+class NormalizeObservation(Wrapper):
+    """Normalize the observation.
+
+    Taken and adapted from https://github.com/luchris429/purejaxrl/blob/main/purejaxrl/wrappers.py
```
Comment (ffelten) [inline crazy_rl/utils/jax_wrappers.py:217]: Comment outdated

**Review by ffelten — COMMENTED** [review]
Almost done :-)

File: crazy_rl/test/test_numpy_and_jax.py (hunk 4 above)
Comment (ColineLedez) [reply crazy_rl/test/test_numpy_and_jax.py:24]: It tests if the jax and numpy reward are the same, with an approximation

Comment (ffelten) [reply crazy_rl/test/test_numpy_and_jax.py:24]: Isn't something like `np.allclose(np.asarray(jax_rewards[0]), np_rewards["agent_0"]))` doable?
